8355585: Aarch64: Add aarch64 backend for Float16 vector operations

Bhavana-Kilambi opened the pull request:
This patch adds aarch64 backend (both Neon and SVE) for FP16 vector operations - add, mul, sub, div, min, max, sqrt and fma.

Testing:
JTREG tests - hotspot_all, jdk (tier 1-3) and langtools (tier 1) pass on aarch64 which also includes the JTREG test to test the FP16 vector operations - `test/hotspot/jtreg/compiler/vectorization/TestFloat16VectorOperations.java`

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8355585](https://bugs.openjdk.org/browse/JDK-8355585): Aarch64: Add aarch64 backend for Float16 vector operations (**Enhancement** - P4)


### Reviewers
 * [Hao Sun](https://openjdk.org/census#haosun) (@shqking - Committer) Review applies to [31103f69](https://git.openjdk.org/jdk/pull/25096/files/31103f695f1d7d30c656f801ad6017960cad23f4)
 * [Xiaohong Gong](https://openjdk.org/census#xgong) (@XiaohongGong - Committer) Review applies to [bb6235aa](https://git.openjdk.org/jdk/pull/25096/files/bb6235aa5b23d73a480237b0119011ada49dca11)
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**) Review applies to [bb6235aa](https://git.openjdk.org/jdk/pull/25096/files/bb6235aa5b23d73a480237b0119011ada49dca11)
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/25096/head:pull/25096` \
`$ git checkout pull/25096`

Update a local copy of the PR: \
`$ git checkout pull/25096` \
`$ git pull https://git.openjdk.org/jdk.git pull/25096/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 25096`

View PR using the GUI difftool: \
`$ git pr show -t 25096`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/25096.diff">https://git.openjdk.org/jdk/pull/25096.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/25096#issuecomment-2858783333)
</details>

shqking commented on src/hotspot/cpu/aarch64/aarch64_vector_ad.m4:
```diff
@@ -215,6 +215,26 @@ source %{
+      case Op_MinVHF:
+      case Op_MaxVHF:
+      case Op_SqrtVHF:
+        // FEAT_FP16 is enabled if both "fphp" and "asimdhp" features are supported.
```
It's an unary op and we should add it to `is_vector_unary_op_name` in `adlc/dfa.cpp`.
See the related code in the previous patch https://github.com/openjdk/jdk/pull/9534

XiaohongGong commented on src/hotspot/cpu/aarch64/assembler_aarch64.hpp:
```diff
@@ -2743,26 +2743,26 @@ template<typename R, typename... Rx>
-  INSN(fmax, 0, 0, 0b111101);
-  INSN(fmin, 0, 1, 0b111101);
-  INSN(facgt, 1, 1, 0b111011);
+  // Advanced SIMD three same
+  void adv_simd_three_same(Instruction_aarch64 &current_insn, FloatRegister Vd,
+                           SIMD_Arrangement T, FloatRegister Vn, FloatRegister Vm,
+                           int op1, int op2, int op3);
```
May I ask why you move this to the .cpp file?

theRealAph replied:
It's not a bad thing to do. Once the function gets so large, especially when it is inlined many times, that's a wise thing to do.

XiaohongGong replied:
Make sense to me. Thanks for your explanation @theRealAph !

XiaohongGong commented on src/hotspot/cpu/aarch64/aarch64_vector.ad:
```diff
@@ -657,6 +693,16 @@ instruct vaddL_masked(vReg dst_src1, vReg src2, pRegGov pg) %{
   ins_pipe(pipe_slow);
 %}
 
+instruct vaddHF_masked(vReg dst_src1, vReg src2, pRegGov pg) %{
+  predicate(UseSVE > 0);
+  match(Set dst_src1 (AddVHF (Binary dst_src1 src2) pg));
```
Do we have such a case in existing jtreg now?

Bhavana-Kilambi replied:
Not at the moment. Thanks for pointing this out. I think for this PR, I will remove the predicated instructions support and for now only keep support for non-masked ones. I will add this support when there will be more focus on the masked versions once VectorAPI with Float16Vector is integrated with mainline. Hope this is ok.

XiaohongGong replied:
Sounds good to me. But I'm worried it may crash with bad ad file on AArch64 if the Vector API java and compiler IR part is ready for HF types, while the AArch64 relative masked rules are missing.  Beacause the masked vector IR have been generated, while the codegen is missing on AArch64. We have to add the HF ops to `match_rule_supported_vector_masked` first, and then remove them when adding the masked version rules. WDYT?

Bhavana-Kilambi replied:
It's a good idea to disable masked ops in the backend. That should help not generate the masked IR in the inline expanders. I'll update this PR soon. Thanks for your comments.

Bhavana-Kilambi replied:
Done. Thanks!

shqking commented on src/hotspot/cpu/aarch64/aarch64_vector_ad.m4:
```diff
@@ -508,20 +541,21 @@ dnl
-BINARY_OP_PREDICATE(vaddL, AddVL, sve_add,  D)
-BINARY_OP_PREDICATE(vaddF, AddVF, sve_fadd, S)
-BINARY_OP_PREDICATE(vaddD, AddVD, sve_fadd, D)
+BINARY_OP_PREDICATE(vaddB,  AddVB,  sve_add,  B)
```
stylistic nit: remove the extra spaces.

In the initial commit, this extra space is added due to `vaddHF`. However, in the latest commit the predicated Float16 rules are removed.

Bhavana-Kilambi replied:
Done. Thanks!

shqking reviewed (commented):
LGTM except one nit.

XiaohongGong reviewed (approved):
LGTM! Thanks!

Bhavana-Kilambi commented:
Hi @shqking , does it look ok to you for an approval now ?

shqking reviewed (approved):
Thanks for your update. LGTM.

Bhavana-Kilambi commented:
Hi @theRealAph , Can I request you to please take a look at this PR? I need a "reviewer" to review my change as well for the PR to be ready for integration. Thanks in advance!

eme64 commented:
@Bhavana-Kilambi I'd like to run some testing before integration, please ping me again in 24h for results :)

XiaohongGong reviewed (approved):
LGTM. Thanks!

Bhavana-Kilambi commented:
> Looks good. I'm assuming you've tested both SVE and Neon.

Yes, this was tested on both SVE and Neon (N1/V1/V2 architectures).

eme64 commented:
@Bhavana-Kilambi I'm getting timeouts with your new test:
`compiler/vectorization/TestFloat16VectorOperations.java`

At least on `linux-aarch64-debug` and `windows-x64-debug`, but not all tests have completed yet, so more could be failing.

Not sure if it is relevant, but both had extra flag `-XX:-UseTLAB`, we add this flag in our additional stress testing.

Bhavana-Kilambi commented:
> @Bhavana-Kilambi I'm getting timeouts with your new test: `compiler/vectorization/TestFloat16VectorOperations.java`
> 
> At least on `linux-aarch64-debug` and `windows-x64-debug`, but not all tests have completed yet, so more could be failing.
> 
> Not sure if it is relevant, but both had extra flag `-XX:-UseTLAB`, we add this flag in our additional stress testing.

Thanks for letting me know. This test by default takes a very long time to finish. As i added more flags (to test various vector sizes) to be tested, it probably ran way too long than anticipated and resulted in the timeout error. I will update with a patch to remove testing these extra flags for now. I did try to increase the default timeout value but the test continued to run for more than 20 min after which I had to terminate. I feel the best for now is to remove the additional tests for various vector sizes that I have added. I will update the patch soon.

eme64 commented:
@Bhavana-Kilambi Ok, yes, 20min is a bit excessive :laughing: 

Generally, we should periodically run all vector tests with various `MaxVectorSize` settings. But doing that all the time is often too time consuming. For some specific tests, it can make sense though to iterate over multiple sizes.

I wonder if you could also reduce the runtime of the test in other ways? Maybe reduce the warmup? It seems a bit excessive to do `10000` warmup iterations, which each execute a loop with many iterations themselves.

Bhavana-Kilambi commented:
Hi @eme64 I removed the @Warmup entirely and the test does pass on aarch64. Although I am a bit afraid to fully remove it as it could sometimes lead to the loop not being warm enough for c2 vectorization to kick in. I haven't tried with different values of the warmup iterations though. Do you think it's ok to remove it entirely?

eme64 commented:
@Bhavana-Kilambi The TestFramework actually forces C2 compilation:
- runs warmup iterations, maybe C2 triggers automatically because there are enough iterations.
- Once warmup is over, the TestFramework checks if the method is already compiled, if not, it enqueues it.
- In the end, we know it is C2 compiled, which gives us the C2 IR we can match with.

In my experience, having low warmup count works in most cases. Except when you need profiling data. If you have zero warup, we basically have compilation with `-Xcomp`.

So it really depends on your specific case. In general, I would avoid doing an `Xcomp` compilation / zero warmup, because then we do not test normal compilation with profiling. And compilation with profiling is more important I think.

But in cases where you have a large loop in the test method, we would trigger OSR and normal compilation with profiling rather soon anyway. So lowering the warmup is ok. How many loop iterations do we need for OSR?
`product(intx, Tier4BackEdgeThreshold, 40000`. We could round that up to `100_000`, just to be sure. With `LEN = 2048`, you would thus only need about `50` invocations of the tests during warmup to reach C2 compilation. Hence, the current `@Warmup(10000)` is much too high, I think. You could cut down the runtime by about a factor of `100` here, if my math is correct :exploding_head: 

What do you think?

Bhavana-Kilambi commented:
Hi @eme64 Thanks for the details and suggestions. 
I tried with a `@Warmpup(50) `(my calculation is 50 * 2048 = 102400 which is around 100_000) and the test passes on aarch64 (it passes even with 0 warmp though). Do you think we can go ahead with `@Warmup(50)` ?

Also, can I ask if any other tests failed on your side (they shouldn't though as I havent touched any other code other than FP16)?

Bhavana-Kilambi commented:
Hi @eme64 I have updated the testcase. Could you please test it now if it works? Thanks!

eme64 commented:
@Bhavana-Kilambi launched! :rocket:

Bhavana-Kilambi commented:
Hi @eme64 Hope the tests have passed !

eme64 reviewed (approved):
Thanks for the work and patience :)

Changes look reasonable, though I did not review them in detail.

Tests have passed!

XiaohongGong commented:
/sponsor
